### PR TITLE
UserList changes, Group fetching logic changes based on changes in keycloak

### PR DESCRIPTION
**Pull request by dinesh-aot**
#1439 
- Included display name and changes in group fetching logic from keycloak
- Corresponding changes in UserList.tsx

### CODE REVIEW

File: epictrack-api/src/api/services/user.py [hunk 1]
```diff
@@ -26,40 +26,67 @@ def get_all_users():
         """Get all users"""
         users = KeycloakService.get_users()
         for user in users:
-            user['group'] = None
+            user["group"] = None
         groups = UserService.get_groups()
-        groups = sorted(groups, key=lambda x: x['attributes']['level'][0])
+        groups = sorted(groups, key=lambda x: x["attributes"]["level"][0])
```
Comment (saravanpa-aot) [inline epictrack-api/src/api/services/user.py]: u can use the get_level() method here?
groups = UserService.get_groups()
sorted_groups = sorted(groups, key=get_level)

File: epictrack-api/src/api/services/user.py [hunk 2]
```diff
@@ -26,40 +26,67 @@ def get_all_users():
...
         groups = UserService.get_groups()
-        groups = sorted(groups, key=lambda x: x['attributes']['level'][0])
+        groups = sorted(groups, key=lambda x: x["attributes"]["level"][0])
         for group in groups:
-            members = KeycloakService.get_group_members(group['id'])
-            member_ids = [member['id'] for member in members]
-            filtered_users = list(filter(lambda x, _member_ids=member_ids: x['id'] in _member_ids, users))
+            members = KeycloakService.get_group_members(group["id"])
+            member_ids = [member["id"] for member in members]
+            filtered_users = list(
```
Comment (saravanpa-aot) [inline epictrack-api/src/api/services/user.py]: filtered_users = [user for user in users if user["id"] in member_ids]

File: epictrack-api/src/api/services/user.py [hunk 3]
```diff
@@ -21,45 +21,73 @@
...
 
-    @staticmethod
-    def get_groups():
+    @classmethod
+    def get_groups(cls):
         """Get groups that has level set up"""
         groups = KeycloakService.get_groups()
-        filtered_groups = list(filter((lambda g: 'level' in g['attributes']), groups))
+        filtered_groups = []
+        for group in groups:
```
Comment (jadmnsaad) [inline epictrack-api/src/api/services/user.py:45]: can we simplify the nesting, and perhaps break this into different functions that makes it easier to understand. Suggestion:

    @classmethod
    def get_groups_with_level(cls):
        """Get groups that have a 'level' attribute set up."""
        all_groups = KeycloakService.get_groups()
        sub_groups_with_level = cls._get_sub_groups_with_level(all_groups)
        groups_with_level = cls._get_groups_with_level(all_groups)

        return sub_groups_with_level + groups_with_level

    @classmethod
    def _get_sub_groups_with_level(cls, all_groups):
        """Get sub-groups with 'level' attribute."""
        return [
            sub_group
            for group in all_groups
            for sub_group in group.get("subGroups", [])
            if "level" in sub_group.get("attributes", {})
        ]

    @classmethod
    def _get_groups_with_level(cls, all_groups):
        """Get groups with 'level' attribute."""
        return [
            group
            for group in all_groups
            if "level" in group.get("attributes", {})
        ]

File: epictrack-api/src/api/services/user.py [hunk 4]
```diff
@@ -21,45 +21,73 @@
...
+                if group["id"] == user_group_request.get("group_id_to_update")
+            ),
+            None,
+        )
+        if (
+            not requested_group
+            and not updating_group
+            and int(requested_group["attributes"]["level"][0])
+            < int(updating_group["attributes"]["level"][0])
+        ):
```
Comment (jadmnsaad) [inline epictrack-api/src/api/services/user.py:77]: can export this part into a _validate_update function. I'm thinking of the Single Responsibility Principle and how we can make this update_user_group function do less

**Pull request merged**